docs/notebooks/trust_region.pct.py

```diff
@@ -0,0 +1,286 @@
+        grid_density=40,
+        contour=True,
+    )
+
+    plot_bo_points(query_points, ax[0, 0], num_initial_data_points, arg_min_idx)
+
+
+plot_final_result(dataset)
+
+# %% [markdown]
+# We can also visualize the progress of the optimization by plotting the trust regions at each step.
+# The trust regions are shown as translucent boxes, with the current optimum point in each region
+# shown in matching color.
+#
+# Note there is only one trust region in this plot, but the rule in the next section will show multiple trust
+# regions.
+
+# %%
+import base64
+import io
+from typing import List
+
+import imageio
+import IPython
+import matplotlib.pyplot as plt
+
+from trieste.experimental.plotting import plot_trust_region_history_2d
+
+
+def fig_to_frame(fig: plt.Figure) -> np.ndarray:
+    fig.canvas.draw()
+    size_pix = fig.get_size_inches() * fig.dpi
+    image = np.frombuffer(fig.canvas.tostring_rgb(), dtype="uint8")
+    return image.reshape(list(size_pix[::-1].astype(int)) + [3])
+
+
+def frames_to_gif(
+    frames: List[np.ndarray], duration=5000
+) -> IPython.display.HTML:
+    gif_file = io.BytesIO()
+    imageio.mimsave(gif_file, frames, format="gif", loop=0, duration=duration)
+    gif = IPython.display.HTML(
+        '<img src="data:image/gif;base64,{0}"/>'.format(
+            base64.b64encode(gif_file.getvalue()).decode()
+        )
+    )
+    return gif
+
+
+def plot_history(result: trieste.bayesian_optimizer.OptimizationResult) -> None:
+    frames = []
+    for step, hist in enumerate(
+        result.history + [result.final_result.unwrap()]
+    ):
+        fig, _ = plot_trust_region_history_2d(
+            branin,
+            search_space.lower,
+            search_space.upper,
+            hist,
+            num_init=num_initial_data_points,
+        )
+
+        if fig is not None:
+            fig.suptitle(f"step number {step}")
+            frames.append(fig_to_frame(fig))
+            plt.close(fig)
+
+    IPython.display.display(frames_to_gif(frames))
+
+
+plot_history(result)
+
+# %% [markdown]
+# ## Batch trust region rule
+#
+# Next we demonstrate how to run Bayesian optimization with the batch trust region rule.
+#
+# ### Create the batch trust region acquisition rule
+#
+# We achieve Bayesian optimization with trust region by specifying `BatchTrustRegionBox` as the
+# acquisition rule.
+#
+# This rule needs an initial number `num_query_points` of sub-spaces (or trust regions) to be
+# provided and performs optimization in parallel across all these sub-spaces. Each region
+# contributes one query point, resulting in each acquisition step collecting `num_query_points`
+# points overall. As the optimization process continues, the bounds of these sub-spaces are
+# dynamically updated.
+#
+# In addition, this rule requires the specification of a batch aquisition base-rule for performing
+# optimization; for our example we use `EfficientGlobalOptimization` coupled with
+# `ParallelContinuousThompsonSampling`.
```

Comment:
suggestion:
... coupled with the `ParallelContinuousThompsonSampling` acquisition function.